### ISSUE TITLE
File processing

### ISSUE THREAD

**Issue by pdb5627** [GridProtectionAlliance/openMIC#9]
I just set up a test installation of openMIC trying to read event files out of an SEL-351 relay over FTP. It looks like openMIC is correctly connecting to the FTP service and enumerates the files in the designated directory. After that, no files are retrieved but no status indication of why is given. Any ideas?
```
[LAB351] Attempting connection to FTP server "Test@xxx.xxx.xxx.xxx:20021"...
[LAB351] FTP Request: USER ************
[LAB351] FTP Response: 331 User name okay, need password.
[LAB351] FTP Request: PASS ************
[LAB351] FTP Response: 230 User logged in, proceed.
[LAB351] FTP Request: PWD
[LAB351] FTP Response: 257 "/" is current directory.
[LAB351] Connected to FTP server "Test@xxx.xxx.xxx.xxx:20021"
[LAB351] Starting "TEST-351" connection profile "SEL-351" task processing:
[LAB351] Attempting to set remote FTP directory path "/EVENTS"...
[LAB351] FTP Request: CWD /EVENTS/
[LAB351] FTP Response: 250 CWD  requested file action okay, completed.
[LAB351] FTP Request: TYPE A
[LAB351] FTP Response: 200 TYPE Command okay.
[LAB351] FTP Request: PASV
[LAB351] FTP Response: 227 Entering Passive Mode (xxx,xxx,xxx,xxx,152,40).
[LAB351] FTP Request: LIST
[LAB351] FTP Response: 125 Data connection already open; transfer starting.
[LAB351] FTP Response: 226 Closing data connection.
[LAB351] Enumerating remote files in "/EVENTS"...
[LAB351] Found 117 remote files, starting file processing...
[LAB351] FTP session connected for 2.07 seconds
```

**Comment by ritchiecarroll** [comment]
May need to check the "Log Detailed Connection Messages" on the device screen -or- you may have your file filter options incorrectly set...

**Comment by pdb5627** [comment]
Log Detailed Connection Messages is enabled, thus you see the back and forth of FTP commands in the log. The connection string:
```
connectionHostName=xxx.xxx.xxx.xxx:20021; connectionUserName=************; connectionPassword=************; connectionProfileID=1; schedule=*/2 * * * *; useDialUp=false; dialUpEntryName=; dialUpNumber=; dialUpUserName=; dialUpPassword=; dialUpRetries=3; dialUpTimeout=90; connectionTimeout=30000; logConnectionMessages=true
```

The connection profile task is set with file name filter `*.*`. The full task settings string:
```
fileExtensions=*.*; remotePath=/EVENTS; localPath=C:\openMIC; recursiveDownload=false; deleteRemoteFilesAfterDownload=false; limitRemoteFileDownloadByAge=false; deleteOldLocalFiles=false; skipDownloadIfUnchanged=true; overwriteExistingLocalFiles=false; archiveExistingFilesBeforeDownload=false; synchronizeTimestamps=true; externalOperation=; maximumFileSize=1000; maximumFileCount=-1; directoryNamingExpression=<YYYY><MM>\<DeviceFolderName>; directoryAuthUserName=; directoryAuthPassword=; emailOnFileUpdate=false; emailRecipients=
```

**Comment by ritchiecarroll** [comment]
That's interesting. Have you tried a manual FTP connection with the same credentials?

**Comment by pdb5627** [comment]
Yep. The FTP login succeeds. The openMIC log above shows it connects over FTP, retrieves the directory listing, and correctly enumerates the number of files before exiting. Looking at the Downloader.cs code, it almost has to be failing the file pattern match, otherwise it would log that it is "Processing remote file ____", right?

**Comment by ritchiecarroll** [comment]
Do any of the files to be downloaded have extensions and/or what do the file names look like?

**Comment by ritchiecarroll** [comment]
You are correct, BTW, it must be in the loop and exiting here:
```
    if (!FilePath.IsFilePatternMatch(settings.FileSpecs, file.Name, true))
        continue;
```
Otherwise there would be a logged exception...

So either there is a bug in `FilePath.IsFilePatternMatch` for the kinds of file names being encountered - or- the `file.Name` being reported is black or something.

**Comment by pdb5627** [comment]
Files are have names like `C4_10210.CEV`, `CHISTORY.TXT`, `CR_10210.CEV`, `HR_10210.CFG`, 'HR_10210.DAT`, `HR_10210.HDR` and so forth.

From a Wireshark capture of the traffic to the relay, the relay responds to the FTP LIST command with the following:
```
-r--r--r--  1 user     group       1781  Mar 17 18:29 CHISTORY.TXT\000\r\n-r--r--r--  1 user     group       1613  Mar 17 18:29 HISTORY.TXT\000\r\n-r--r--r--  1 user     group      60416  Dec 19 17:16 C4_10232.CEV\000\r\n-r--r--r
```
and so forth.

**Comment by ritchiecarroll** [comment]
It could be that our FTP engine is having trouble parsing those file names...

We will manually try the string provided and see what comes out.

Thanks,
Ritchie

**Comment by pdb5627** [comment]
With that null character after each filename before the line feed & character return, I wonder if the MatchingListLine function in FtpDirectory.cs thinks the null character is part of the filename.

**Comment by baites1** [comment]
Hi @ritchiecarroll, I'm wondering if you guys had any luck resolving this issue?

I too am trying to use openMIC to automatically download event files from a SEL-351 relay via FTP. 
I can successfully use Windows explorer to view the contents of each folder on the SEL FTP server, and up to a point, openMIC will download the single "CFG.txt" file in the root directory of the FTP server, but fails to download anything in any of the sub directories. 
I'm using the openMIC WebGUI to configure, and the response I get is "Failed to build file list for remote directory "EVENTS": Cannot change directory to //EVENTS/." Screen shots attached.

![1 - ftp root directory](https://user-images.githubusercontent.com/38897352/41326538-7dd0829a-6f13-11e8-812d-ca0090bd4ec4.png)

![2 - ftp events directory](https://user-images.githubusercontent.com/38897352/41326551-87020e88-6f13-11e8-8bd3-40873d2455f2.jpg)

![3 - openmic status](https://user-images.githubusercontent.com/38897352/41326553-8a53482c-6f13-11e8-8ccf-56fe572f69e1.png)

![4 - openmic task definition](https://user-images.githubusercontent.com/38897352/41326555-8d017404-6f13-11e8-8c70-b51f86c9895a.png)

**Comment by StephenCWills** [comment]
Hi, @baites1. I don't think we've addressed this issue, but I believe the issue you are having has been fixed in the latest codebase. I would suggest downloading the latest nightly build version of openMIC and see if the issue persists.

http://www.gridprotectionalliance.org/NightlyBuilds/openMIC/Beta/openMIC.Installs.zip

**Comment by StephenCWills** [comment]
As for the originally reported issue, I added some code to the FTP client used by openMIC to handle null-terminated filenames in directory listings. Here is the link to the fix.
https://github.com/GridProtectionAlliance/gsf/commit/eac932daf2c892aa88b29e8a0adce2ae44e93e8d

That means this issue should be resolved in tomorrow's nightly build (v1.3.10+). Given the age of this issue, I am going to close it for now. Feel free to reopen the issue if the null terminator continues to cause problems.
